LaTeX/syntax_test_latex.tex

```diff
@@ -28,31 +28,120 @@
 % <- comment.line.percentage.tex
 
 \newcommand{\foo}{\bar}
-%   ^ meta.function.newcommand.latex
-%             ^support.function.latex entity.name.newcommand.latex
+%^^^^^^^^^^ keyword.declaration.function.latex
+%^^^^^^^^^^^^^^^^ meta.function 
+%           ^^^^ entity.name.newcommand.latex
+%           ^ punctuation.definition.backslash.latex
+%                ^^^^^^ meta.function.definition.latex
 %                   ^ support.function.general.latex
 
 \newcommand{\foo}[1]{\bar #1}
-%   ^ meta.function.newcommand.latex
-%             ^support.function.latex entity.name.newcommand.latex
+%   ^ meta.function.latex
+%           ^^^^ entity.name.newcommand.latex
+%                 ^ constant.numeric.value.latex
 %                      ^ support.function.general.latex
 
 \newcommand{\foo}[2][default]{\bar #1 #2}
-%   ^ meta.function.newcommand.latex
-%             ^support.function.latex entity.name.newcommand.latex
+%   ^ meta.function.latex
+%           ^^^^ entity.name.newcommand.latex
+%                 ^ constant.numeric.value.latex
+%                     ^^^^^^ meta.function.parameters.default-value.latex
 %                               ^ support.function.general.latex
 
 
 \renewcommand{\foo}[1]{\bar #1}
-%   ^ meta.function.newcommand.latex
-%               ^support.function.latex entity.name.newcommand.latex
+%   ^ meta.function.latex
+%             ^^^^ entity.name.newcommand.latex
 %                        ^ support.function.general.latex
 
-\providecommand{\foo}[2][default]{\bar #1 #2}
-%   ^ meta.function.newcommand.latex
-%                 ^support.function.latex entity.name.newcommand.latex
-%                                  ^ support.function.general.latex
+  \providecommand{\f@o}[2][default]{\bar #1 #2}
+% ^^^^^^^^^^^^^^^^^^^^^ meta.function.latex
+%                 ^^^^ entity.name.newcommand.latex
+%                      ^^^ meta.function.parameters
+%                         ^^^^^^^^^ meta.function.parameters.default-value.latex
+%                                  ^^^^^^^^^^^^ text.tex.latex
+
+\newcommand\foo{\bar}
+%   ^ meta.function.latex
+%          ^^^^ entity.name.newcommand.latex
+%              ^^^^^^ meta.function.definition.latex - meta.function.latex
+%                 ^ support.function.general.latex
+
+\newcommand{ \foo }{\bar}
+%   ^ meta.function.latex
+%            ^^^^ entity.name.newcommand.latex
+%                     ^ support.function.general.latex
+
+\newcommand* {\foo }{\bar}
+%   ^ meta.function.latex
+%             ^^^^ entity.name.newcommand.latex
+%                      ^ support.function.general.latex
+
+\newcommand \foo {\bar}
+%   ^ meta.function.latex
```

Comment:
If spaces are allowed, we should ensure to maintain an uninterrupted `meta.function` scope covering the whole statement.

```suggestion
\newcommand \foo {\bar}
%^^^^^^^^^^^^^^^^^^^^^^ meta.function
```

This is mainly achieved by `meta_scope` to also cover intermediate comments.

Reply:
The meta scope is probably the most tricky part - as in other syntaxes as well. You are right with stacking `meta.function` not being desirable even though some older syntaxes still do something like that.

The bracket contexts could use `clear_scope: 1` to remove a possibly global `meta.function.latex` meta scope. The `newcommand-definition` context may require special treatment for the parts not surrounded by brackets - in case `meta.function.definition` should keep as a special scope.

If it denotes the function body, it could also be just scoped `meta.function.latex`. This is what most other syntaxes do as well. They use `meta.function meta.block` most likely.

Reply:
thanks, I wasn't really aware of `clear_scope`. That certainly makes things easier. Yes, I think the part in braces generally corresponds to a function body. Though you could also interpret it like e.g. the definition of a macro in the C family of languages. 

In any case, I'm not really sure how to implement this in the syntax file. Since we are pushing multiple contexts, I cannot set a meta_scope there, right?

Reply:
Basically there are 2 major ways of adding seperate meta scopes for each sub-context.

1. push the first context and than set all the others one after another using lookahead patterns. Each of them can then have its own meta scope without the previous one overlapping.
2. use a multi-push statement (as you did here) and add a common scope meta scope to the first one being pushed (the last remaining on stack; `newcommand-definition`) and clear it whenever needed in the other contexts.

The latter is the somewhat more "modern" approach as chances are highier to get re-usable contexts when designing multi-push patterns. Results are more or less the same performance-wise.

I'll try to work out a suggestion how to do that.

Reply:
A suggestion is posted at https://github.com/ngc92/SublimeTextPackages/pull/1

Reply:
@deathaxe Thank you very much. One question, though: You have removed the '' around the match patterns.
Personally, I'm not really a fan of that, because it leads to inconsistencies, e.g. the `immediately-pop` context still uses `'`.
Is this a style convention that you have decided to use for all syntax files? If that is the case, then all the other `match` patterns within the `TeX` and `LaTeX` syntaxes should also be updated, otherwise we have an inconsistent style within the file itself.

Reply:
Over time a habit/coding-style has evolved which includes removing as many quotation/punctuation as possible. It's not a strong language convention though.

While most patterns in LaTeX still use single quotes which may be a result of automatic convertion from tmLanguage format most of them are not really needed.

We could also quote scope names or includes, but we don't do that. The same may apply to patterns. Sure, some may need extra escaping or still require quotes, but it doesn't really justify to quote everything.

Just have a look at other syntaxes, like Java, PHP, ... which have been recently rewritten. Quotes are almost completely gone.

This change could also be handled in a separate PR, if it feels better to you.

Reply:
I think its fine to do it in this PR then, so long as we immediately afterwards have a second PR that strips the other quotes to make the files' style internally consistent again.

Reply:
I would appriciate that.